segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableIterator.java

```diff
@@ -0,0 +1,321 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.tables;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.AsyncIterator;
+import io.pravega.segmentstore.contracts.Attributes;
+import io.pravega.segmentstore.server.AttributeIterator;
+import io.pravega.segmentstore.server.DirectSegmentAccess;
+import java.time.Duration;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.AccessLevel;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+
+/**
+ * Iterates through {@link TableBucket}s in a Segment.
+ * @param <T> Type of the final, converted result.
+ */
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@ThreadSafe
+class TableIterator<T> implements AsyncIterator<T> {
+    //region Members
+
+    private final AttributeIterator indexHashIterator;
```

Comment:
Is this iterating over a snapshot of the attributes or is it changing with updates to the attributes?

Reply:
Every instance of an `AttributeIterator` takes a snapshot of the attributes that are currently cached in the Segment's Metadata. It then creates a `BTreeIndex.iterator` which, as long as we don't truncate the index while iterating, will always provide a consistent iterator (since it keeps pointers to the pages it iterates over and from there it can find their siblings). If the BTreeIndex is truncated due to compaction and the truncation point is beyond the current page, then the BTreeIndex will reinitialize using the latest state and resume the iteration from the value it left off.

So the answer to your question is "yes, in most cases".

The alternative to these would be:
1. Lock the index for writing while iterating. 
2. Prevent compaction while iterating. 

I consider both of these unacceptable because we may have many iterators, concurrent or not. Since iterators are long-lived compared to a single operation, this would introduce unwanted bottlenecks into the functioning of the index.

Reply:
Yes it can be documented.